src/hwperfmon.c

```diff
@@ -0,0 +1,356 @@
+        __asm__ __volatile__("csrr %0, mhpmevent" #x : "=r"(val));             \
+        break;
+
+/* Macro to read HW performance monitor counter values */
+#if __riscv_xlen == 32
+#define METAL_HWPERFMON_GET_COUNT_REG(x)                                       \
+    case METAL_HWPERFMON_COUNTER_##x:                                          \
+        __asm__ __volatile__("csrr %0, mhpmcounter" #x "h" : "=r"(vh));        \
+        __asm__ __volatile__("csrr %0, mhpmcounter" #x : "=r"(vl));            \
+        break;
+#else
+#define METAL_HWPERFMON_GET_COUNT_REG(x)                                       \
+    case METAL_HWPERFMON_COUNTER_##x:                                          \
+        __asm__ __volatile__("csrr %0, mhpmcounter" #x : "=r"(vl));            \
+        break;
+#endif
+
+/* Macro to clear HW performance monitor counter values */
+#if __riscv_xlen == 32
+#define METAL_HWPERFMON_CLR_COUNT_REG(x)                                       \
+    case METAL_HWPERFMON_COUNTER_##x:                                          \
+        __asm__ __volatile__("csrw mhpmcounter" #x "h, zero");                 \
+        __asm__ __volatile__("csrw mhpmcounter" #x ", zero");                  \
+        break;
+#else
+#define METAL_HWPERFMON_CLR_COUNT_REG(x)                                       \
+    case METAL_HWPERFMON_COUNTER_##x:                                          \
+        __asm__ __volatile__("csrw mhpmcounter" #x ", zero");                  \
+        break;
+#endif
+
+/* Macro to check for instruction trap */
+#define MCAUSE_ILLEGAL_INST 0x02
+
+/* Return codes */
+#define METAL_HWPERFMON_RET_OK 0
+#define METAL_HWPERFMON_RET_NOK 1
+
+/* Module enable flag */
+static int hwperfmon_enable = 0;
+
+/* Available hardware counters */
+static metal_hwperfmon_counter hwperfmon_count = 0;
+
+/* Atomic access flag, place it into ITIM. */
+static int flag __attribute__((aligned(4))) __attribute__((section(".itim"))) =
+    0;
+
+__attribute__((naked)) static int metal_hwperfmon_cas(int *ptr, int old,
+                                                      int new) {
+    /* Perform atomic compare and swap */
+#ifdef __riscv_atomic
+    __asm__ __volatile__("amoswap.w.aq t0, t0, (a0) \n\t"
+                         "bne t0, a1, fail # Doesn�t match, so fail. \n\t"
+                         "amoswap.w.rl a2, a2, (a0) \n\t"
+                         "li a0, 0 # Set return to success. \n\t"
+                         "jr ra # Return. \n\t"
+                         "fail: \n\t"
+                         "amoswap.w.rl t0, t0, (a0) \n\t"
+                         "li a0, 1 # Set return to failure. \n\t"
+                         "jr ra # Return. \n\t");
+#endif
+    /* Return error, if no support */
+    __asm__ __volatile__("li a0, 1 \n\t jr ra \n\t");
+    return 1;
+}
+
+int metal_hwperfmon_enable(void) {
+    /* Make sure only a single hart executes this code */

```

Comment:
Do we need to enable atomics on all harts, or is it sufficient to enable them on a single hart?

Reply:
Fixed. All harts can enable their counters.